graylog2-server/src/main/java/org/graylog/plugins/map/config/S3GeoIpFileService.java

```diff
@@ -0,0 +1,315 @@
+    private Instant tempCityFileLastModified = null;
+
+    @Inject
+    public S3GeoIpFileService(@Named(GeoIpProcessorConfig.S3_DOWNLOAD_LOCATION) String s3DownloadLocation) {
+        try {
+            final S3ClientBuilder clientBuilder = S3Client.builder();
+            clientBuilder.credentialsProvider(DefaultCredentialsProvider.create());
+            this.s3Client = clientBuilder.build();
+        } catch (Exception e) {
+            this.s3Client = null;
+            LOG.warn(NULL_S3_CLIENT_MESSAGE);
+            LOG.debug("If not trying to use the Geo Location Processor S3 file refresh feature, the following error can safely be ignored.\n\tERROR : {}", e.getMessage());
+        }
+        this.asnPath = Paths.get(s3DownloadLocation, S3GeoIpFileService.ACTIVE_ASN_FILE);
+        this.cityPath = Paths.get(s3DownloadLocation, S3GeoIpFileService.ACTIVE_CITY_FILE);
+        this.tempAsnPath = Paths.get(s3DownloadLocation, S3GeoIpFileService.TEMP_ASN_FILE);
+        this.tempCityPath = Paths.get(s3DownloadLocation, S3GeoIpFileService.TEMP_CITY_FILE);
+    }
+
+    /**
+     * Downloads the Geo Processor city and ASN database files to a temporary location so that they can be validated
+     *
+     * @param config current Geo Location Processor configuration
+     * @throws S3DownloadException if the files fail to be downloaded
+     */
+    public void downloadFilesToTempLocation(GeoIpResolverConfig config) throws S3DownloadException {
+        if (s3ClientIsNull()) {
+            return;
+        }
+
+        try {
+            cleanupTempFiles();
+            BucketsAndKeys bucketsAndKeys = getBucketsAndKeys(config);
+            GetObjectResponse cityResponse = s3Client.getObject(GetObjectRequest.builder()
+                    .bucket(bucketsAndKeys.cityBucket())
+                    .key(bucketsAndKeys.cityKey()).build(), tempCityPath);
+            setFilePermissions(tempCityPath);
+            tempCityFileLastModified = cityResponse.lastModified();
+
+            if (!config.asnDbPath().isEmpty()) {
+                GetObjectResponse asnResponse = s3Client.getObject(GetObjectRequest.builder()
+                        .bucket(bucketsAndKeys.asnBucket())
+                        .key(bucketsAndKeys.asnKey()).build(), tempAsnPath);
+                setFilePermissions(tempAsnPath);
+                tempAsnFileLastModified = asnResponse.lastModified();
+            }
+        } catch (Exception e) {
+            LOG.error("Failed to retrieve S3 files. {}", e.toString());
+            cleanupTempFiles();
+            throw new S3DownloadException(e.getMessage());
+        }
+    }
+
+    /**
+     * Checks to see if either the ASN or city database file has been updated since the last refresh
+     *
+     * @param config current Geo Location Processor configuration
+     * @return true if the files in S3 have been modified since they were last synced
+     */
+    public boolean checkForNewFilesInS3(GeoIpResolverConfig config) {
```

Comment:
While it is going to happen each time the server restarts, it should be happening it it's own thread and not slowing anything down.

I originally didn't do a comparison to the file on disk because in my head there was a scenario where an older file in S3 didn't mean that the file on disk was the most recent... however, when I tried to put that scenario into words I couldn't do it and now it sounds like that would be a good check. Let me take a stab at adding that check for the initial running of the service and I'll ping you when it is up.

Reply:
Ah yeah, it is a different thread. I think I was perceiving a slowdown because of where I had my breakpoints set up. 
It's definitely much less of a big deal then but still nice if you get that initial check working.

Reply:
I still like the idea. I don't know what scenario I had in my head, but I just added some code to set the `FileLastModified` variables based on the last modified time of the files on disk (if they exist) on server startup. See if that works how you'd expect.

Reply:
Works like a charm!

Reply:
Good deal. I've got a recent commit focused on multi-node environments that I need to test out (and will have you do the same). Once we get that straightened out this should be in good shape.